### PR TITLE
#177638827 update `/property/search` endpoint to use filters and sorting

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Update `/property/search` endpoint to use filters and sorting

#### Description of Task to be completed?
Update `/property/search` endpoint to use filters and sorting

#### How should this be manually tested?
`GET /api/v1/property/search`

#### What are the relevant pivotal tracker stories?
#177638827

### CODE REVIEW

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. The initial startup looks good to me, but I am more interested in the tests. We would need to be more comprehensive with our tests here because this is a very important part of the application. You would need to add tests for these sections and ensure they are comprehensive (combine them with existing sections)

1. Searched Properties - Ensure that the right property is returned with different search terms. This should also include ranged tests like price greater than/ less than a specific value.
2. Pagination - Normal pagination helper class should be enough here
3. Sorting: Ensure that most of the entries work as expected when sorted
4. Filters: Ensure that property can be filtered

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. This looks good and the tests looks great too 👍🏽 .

However, can we add the following tests too:

1. Tests to ensure that sorting works as expected for a searched property. At least, try sorting by createdAt (necessary to sort old and newly added property), price (this is the man of the match), and maybe state.
2. Tests to ensure that searching, filtering, and sorting work together with a single request. That is, the test should search for a property, add filters, and sort parameters to confirm that this would work too.
3. Try testing for ranges too for numeric values like price, bathroom, bedrooms, toilets. At least, add a test to confirm that a user can search for houses from the range of 30 million to 50 million to confirm that this works as expected.

Thank you for the hard work.

File: test/controllers/property.controller.test.js [hunk 1]
```diff
@@ -1883,310 +1883,301 @@ describe('Property Controller', () => {
...
-              expect(res.body.properties[0]._id).to.be.eql(property1._id.toString());
-              expect(res.body.properties[0].assignedUsers).to.not.include(
-                regularUser._id.toString(),
-              );
-              expectResponseToExcludeSensitiveUserData(res.body.properties[0].assignedUsers);
-              expectResponseToExcludeSensitiveUserData(res.body.properties[0].vendorInfo);
-              expectResponseToContainNecessaryVendorData(res.body.properties[0].vendorInfo);
+        [adminUser, vendorUser, regularUser].map((user) =>
+          itReturnsTheRightPaginationValue({
+            endpoint,
```
[the remark this answers is not in the document]
Comment (tobieeniafe) [reply test/controllers/property.controller.test.js:1988]: This break the function tests, such that we'd have queries like `/api/v1/property/search?country=a?limit=1`

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. This looks good to me. I dropped a few comments (mostly nits on using a better test description). The pending stuff is adding filtering by VendorId with test. After that, we are good to go 💯 .

File: server/helpers/filters.js [hunk 2]
```diff
@@ -296,3 +296,14 @@ export const OFFLINE_PAYMENT_FILTERS = {
...
+
+export const SEARCH_FILTERS = {
+  ...ADDRESS_FILTERS,
+  bathrooms: { type: FILTER_TYPE.INTEGER },
+  bedrooms: { type: FILTER_TYPE.INTEGER },
+  features: { type: FILTER_TYPE.ARRAY },
+  houseType: { type: FILTER_TYPE.STRING },
+  price: { type: FILTER_TYPE.INTEGER },
+  toilets: { type: FILTER_TYPE.INTEGER },
+  units: { type: FILTER_TYPE.INTEGER },
```
[the remark this answers is not in the document]
Comment (tobieeniafe) [reply server/helpers/filters.js:309]: No, not yet. The only feasible solution I could come up with is, editing the property model and adding the vendor's name as a field.

[the remark this answers is not in the document]
Comment (andela-hpopoola) [reply server/helpers/filters.js:309]: Not totally sure what you mean by this, We already have the vendorId, please let's use that. I was speaking from the frontend angle (apologies), I should have been more explicit to say that we should allow a user to search by `vendorId `(which is addedBy in the property model)

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for making the changes. Still waiting for you to add the `vendorId` stuff (with tests)  as commented here https://github.com/Highrachy/ballers-api/pull/149#discussion_r612137831 and fix failing tests (I didn't look at why it is failing though) so that I can approve.

File: test/controllers/property.controller.test.js [hunk 3]
```diff
@@ -2002,7 +2002,7 @@ describe('Property Controller', () => {
             await assignPropertyToUser({
               userId: regularUser._id,
               propertyId: property._id,
-              vendor: vendorUser,
+              vendor: invalidVendorUser,
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:2005]: I think the best thing is to use a vendor's Id here, maybe a new vendor's Id. Using an `invalidVendorUser` here makes it seem like we can create vendor with invalid vendors 🤷🏽‍♂️

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. This looks good to me. I left a single non-blocking comment otherwise, this is ready to be merged 💯

**Pull request merged**